### DISCUSSION TITLE
Xamarin.Essentials.AppInfo throws System.TypeInitializationException

### DISCUSSION THREAD

**Discussion by Yveske1966** [AvaloniaUI/Avalonia#14007] [Q&A] [answered]
Debugging shows "This functionality is not implemented in the portable version of this assembly. You should reference the NuGet package from your main application project in order to reference the platform-specific implementation."

Online documentation mentions Platform.Init for Android, but I'm compiling for Desktop (Windows).
Furthermore Platform isn't available in CodeCompletion.

`
using System;
using System.IO;
using System.ComponentModel.DataAnnotations;
using Xamarin.Essentials;

namespace AvaloniaApplication1;

public static class Constants
{ 
    public static string AppName = AppInfo.Name;
    public static string AppVersion = AppInfo.VersionString;
    public static string DatabaseFilename = AppName + ".db3";
    public static string DatabasePath =>
        Path.Combine (Environment.GetFolderPath (Environment.SpecialFolder.MyDocuments), AppName);
}
`

**Comment by timunie** [comment, marked as answer]
AppInfo isn't needed for Desktop. There are build in ways to get the needed info. https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assemblyname?view=net-8.0

**Reply by Yveske1966** [reply]
This code worked for Maui (which has Essentials build in), so I'm wondering what is missing to make it work ?
I would also like to use other members of the Essentials package, so your solution is just going to delay the inevitable.

For future reference: Would your solution give me problems when I'm targeting all platforms in the near future, including Raspberry Pi ?
Remark: Using Assembly throws a "cannot declare instance members in a static class".

**Reply by timunie** [reply]
You can for sure use platform specific code. However if that worked for Maui on Windows I suppose it should also work for Avalonia. Maybe you have to do what the message says and add the package to your project explicitly? Can you check this?

**Reply by thevortexcloud** [reply]
>For future reference: Would your solution give me problems when I'm targeting all platforms in the near future, including Raspberry Pi ?

Linux in ,NET uses the standard desktop runtime which is also used by Mac and Windows, so it should work fine. Everything else however I can't say.

> Remark: Using Assembly throws a "cannot declare instance members in a static class".

How are you referencing the assembly? EG `Assembly.GetExecutingAssembly()` is a static method and should work fine.

https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getexecutingassembly?view=net-8.0

Did you forget to make your variable static?

On a side note, despite your class's name, those are in fact not constants. For primitive types (and strings) you can use `const` to make them real constants. For everything else you can use `readonly`

**Reply by Yveske1966** [reply]
To be fair, your solution works !
Thanks !

**Comment by thevortexcloud** [comment]
On a somewhat related tangent: Have you tried Avalonia Essentials? It's a port of Xamarin Essentials but for Avalonia. You do however need to build it from source currently as it is not considered release ready yet. But it looks like `AppInfo` has some implementations.

https://github.com/AvaloniaUI/Avalonia.Essentials
